[FLINK-32028][connectors/elasticsearch] Allow customising bulk failure handling

schulzp opened the pull request:
Extracted `BulkResponseInspector` interface to allow custom handling of (partially) failed bulk requests. If not overridden, default behaviour remains unchanged and partial failures are escalated.

* fixes https://issues.apache.org/jira/browse/FLINK-32028
* allows custom metrics to be exposed

reta commented:
@schulzp there was a similar change introduced into `flink-connector-opensearch`, I believe we could backport it to the Elasticsearch connector to have a similar model of configuration and failure handling.

[1] https://github.com/apache/flink-connector-opensearch/pull/11

reta commented:
> @reta, I checked out that the opensearch failure handler. That achieves the same in regard of error handling. However, it does not allow capturing additional metrics. At least for us that would not suffice.

Thanks @schulzp , I think we could:
 - backport the change from Opensearch connector
 - enhance it with an ability to capture additional metrics
 - port this change to Opensearch and Elasticsearch connectors

reta commented:
> I assume there shared code to be reused and that is more about API consistency?

There is no shared code (I think you meant that) and indeed, it is more about API consistency for both connectors

afedulov commented:
@schulzp thanks! 
We try to avoid Mockito usage, unless it is not possible because because of external dependencies that required concrete classes rather than interfaces. Generally speaking, resorting to Mockito is discouraged.
https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#avoid-mockito---use-reusable-test-implementations
In this case it is not necessary, you can just create mock test implementations as private static classes within the test.

schulzp commented:
@afedulov, sure, I'll implement it with plain java.

schulzp commented:
@afedulov, fixed the test (and extended it)

@reta, I added a `FailureHandler` with a default implementation that resembles the current fail-on-any-failure behavior.

schulzp commented on flink-connector-elasticsearch-base/src/test/java/org/apache/flink/connector/elasticsearch/sink/DefaultBulkResponseInspectorTest.java:
```diff
@@ -0,0 +1,119 @@
+                        })
+                .doesNotThrowAnyException();
+    }
+}
```
I went with `@VisibleForTesting` to expose the relevant fields.

schulzp commented:
@reswqa, fixed the cause of the `NotSerializableException` (implicit reference of anonymous class to `ElasticsearchSinkBuilderBase`). CI should pass now.

schulzp commented:
@reswqa, there are two things that need to be fixed:

1. use `MetricsGroupTestUtils#mockWriterMetricGroup()` instead of `InternalSinkWriterMetricGroup.mock()`, however, the underlying change was introduced [only recently](https://github.com/apache/flink/commit/92951a05127f1e0e2ab0ea04ae022659fc5276ab) code relying on it would not be compatible with 1.17.1 or 1.18.0
2. fix usage of non-public API which is enforced by the `ConnectorRules`, but this would go beyond the scope of this PR, since the main branch triggers those violations (only Java 11 and 17) already, see below

```
[ERROR] Failures: 
[ERROR]   Architecture Violation [Priority: MEDIUM] - Rule 'Connector production code must depend only on public API when outside of connector packages' was violated (11 times):
Constructor <org.apache.flink.connector.elasticsearch.sink.ElasticsearchSink.<init>(java.util.List, org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter, org.apache.flink.connector.base.DeliveryGuarantee, org.apache.flink.connector.elasticsearch.sink.BulkProcessorBuilderFactory, org.apache.flink.connector.elasticsearch.sink.BulkProcessorConfig, org.apache.flink.connector.elasticsearch.sink.NetworkClientConfig)> calls method <org.apache.flink.util.Preconditions.checkArgument(boolean, java.lang.Object)> in (ElasticsearchSink.java:73)
Constructor <org.apache.flink.connector.elasticsearch.sink.ElasticsearchSink.<init>(java.util.List, org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter, org.apache.flink.connector.base.DeliveryGuarantee, org.apache.flink.connector.elasticsearch.sink.BulkProcessorBuilderFactory, org.apache.flink.connector.elasticsearch.sink.BulkProcessorConfig, org.apache.flink.connector.elasticsearch.sink.NetworkClientConfig)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object)> in (ElasticsearchSink.java:71)
Constructor <org.apache.flink.connector.elasticsearch.sink.ElasticsearchSink.<init>(java.util.List, org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter, org.apache.flink.connector.base.DeliveryGuarantee, org.apache.flink.connector.elasticsearch.sink.BulkProcessorBuilderFactory, org.apache.flink.connector.elasticsearch.sink.BulkProcessorConfig, org.apache.flink.connector.elasticsearch.sink.NetworkClientConfig)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object)> in (ElasticsearchSink.java:72)
Constructor <org.apache.flink.connector.elasticsearch.sink.ElasticsearchSink.<init>(java.util.List, org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter, org.apache.flink.connector.base.DeliveryGuarantee, org.apache.flink.connector.elasticsearch.sink.BulkProcessorBuilderFactory, org.apache.flink.connector.elasticsearch.sink.BulkProcessorConfig, org.apache.flink.connector.elasticsearch.sink.NetworkClientConfig)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object)> in (ElasticsearchSink.java:74)
Constructor <org.apache.flink.connector.elasticsearch.sink.ElasticsearchSink.<init>(java.util.List, org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter, org.apache.flink.connector.base.DeliveryGuarantee, org.apache.flink.connector.elasticsearch.sink.BulkProcessorBuilderFactory, org.apache.flink.connector.elasticsearch.sink.BulkProcessorConfig, org.apache.flink.connector.elasticsearch.sink.NetworkClientConfig)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object)> in (ElasticsearchSink.java:75)
Constructor <org.apache.flink.connector.elasticsearch.sink.ElasticsearchSink.<init>(java.util.List, org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter, org.apache.flink.connector.base.DeliveryGuarantee, org.apache.flink.connector.elasticsearch.sink.BulkProcessorBuilderFactory, org.apache.flink.connector.elasticsearch.sink.BulkProcessorConfig, org.apache.flink.connector.elasticsearch.sink.NetworkClientConfig)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object)> in (ElasticsearchSink.java:76)
Constructor <org.apache.flink.connector.elasticsearch.sink.ElasticsearchSink.<init>(java.util.List, org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter, org.apache.flink.connector.base.DeliveryGuarantee, org.apache.flink.connector.elasticsearch.sink.BulkProcessorBuilderFactory, org.apache.flink.connector.elasticsearch.sink.BulkProcessorConfig, org.apache.flink.connector.elasticsearch.sink.NetworkClientConfig)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object)> in (ElasticsearchSink.java:77)
Constructor <org.apache.flink.connector.elasticsearch.sink.ElasticsearchWriter.<init>(java.util.List, org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter, boolean, org.apache.flink.connector.elasticsearch.sink.BulkProcessorConfig, org.apache.flink.connector.elasticsearch.sink.BulkProcessorBuilderFactory, org.apache.flink.connector.elasticsearch.sink.NetworkClientConfig, org.apache.flink.metrics.groups.SinkWriterMetricGroup, org.apache.flink.api.common.operators.MailboxExecutor)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object)> in (ElasticsearchWriter.java:107)
Constructor <org.apache.flink.connector.elasticsearch.sink.ElasticsearchWriter.<init>(java.util.List, org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter, boolean, org.apache.flink.connector.elasticsearch.sink.BulkProcessorConfig, org.apache.flink.connector.elasticsearch.sink.BulkProcessorBuilderFactory, org.apache.flink.connector.elasticsearch.sink.NetworkClientConfig, org.apache.flink.metrics.groups.SinkWriterMetricGroup, org.apache.flink.api.common.operators.MailboxExecutor)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object)> in (ElasticsearchWriter.java:97)
Constructor <org.apache.flink.connector.elasticsearch.sink.ElasticsearchWriter.<init>(java.util.List, org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter, boolean, org.apache.flink.connector.elasticsearch.sink.BulkProcessorConfig, org.apache.flink.connector.elasticsearch.sink.BulkProcessorBuilderFactory, org.apache.flink.connector.elasticsearch.sink.NetworkClientConfig, org.apache.flink.metrics.groups.SinkWriterMetricGroup, org.apache.flink.api.common.operators.MailboxExecutor)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object)> in (ElasticsearchWriter.java:99)
Method <org.apache.flink.connector.elasticsearch.sink.ElasticsearchWriter.extractFailures(org.elasticsearch.action.bulk.BulkRequest, org.elasticsearch.action.bulk.BulkResponse)> calls method <org.apache.flink.util.ExceptionUtils.firstOrSuppressed(java.lang.Throwable, java.lang.Throwable)> in (ElasticsearchWriter.java:264)
```

schulzp commented:
@reswqa, thanks `TestingSinkWriterMetricGroup` works on all tested versions. I only faced the arch rules issue locally, so far.
